### PR TITLE
#161779792 List users

### PR DESCRIPTION
**Pull request by nwashangai**
#### What does this PR do?
* Adds a list Users functionality
#### Description of Task to be completed?
* Unit test to test list user(authors) endpoints
* Create endpoint to get all authors
* Create endpoint to get all authors that matches search
#### How should this be manually tested?
- npm run migrate
- Run npm test
#### Any background context you want to provide?
* N/A
#### What are the relevant pivotal tracker stories?
*  [ #161779792](https://www.pivotaltracker.com/story/show/161779792)
#### Screenshots (if appropriate)
* N/A
#### Questions:   
* N/A

### CODE REVIEW

File: src/controller/userController.js [hunk 1]
```diff
@@ -47,6 +48,77 @@ class UserController {
...
+            'image',
+            'bio',
+            'username'
+          ],
+          where: {
+            isVerified: true,
+          }
+        });
+      }
+      res.status(200).json({
```
Comment (Daymorelah) [inline src/controller/userController.js]: You response message, to me, is not self-explanatory. Something like "Request completed successfully" looks more explanatory.

**Review by Daymorelah — COMMENTED** [review]
Good job @nwashangai!

File: src/controller/userController.js [hunk 2]
```diff
@@ -47,6 +48,77 @@ class UserController {
...
+        message: 'successfully',
+        authors,
+      });
```
Comment (Daymorelah) [inline src/controller/userController.js:107]: Your response message, 'successfully', on line 112 is not self-explanatory. I think something like 'Request completed successfully' is more self-explanatory.

File: src/models/user.js [hunk 3]
```diff
@@ -59,7 +59,7 @@ export default (sequelize, DataTypes) => {
       type: DataTypes.TEXT,
       validate: {
         is: {
-          args: /^[A-Za-z]+[A-Za-z0-9 _.,!?;"']+$/i,
+          args: /^[A-Z]+[A-Z0-9 _\.\-,!?;"']+$/i,
```
Comment (Daymorelah) [inline src/models/user.js]: " ? " is a special character in Regular Expression patterns. You will need to escape it with " \ "

File: src/controller/userController.js (hunk 2 above)
Comment (kingisaac95) [reply src/controller/userController.js:107]:
```
res.status(200).json({
        success: true,
        ...
});
```

File: src/controller/userController.js [hunk 4]
```diff
@@ -47,6 +48,77 @@ class UserController {
...
+  * Route: GET: /authors/:search
+  * @param {object} req -Request object
+  * @param {object} res - Response object
+  * @return {res} res - Response object
+  * @memberof USerController
+ */
+  static async getAuthors(req, res) {
+    try {
+      let authors;
+      if (req.query.search) {
```
Comment (kingisaac95) [inline src/controller/userController.js:80]:
```
const findAllAuthors = async (req) => {
  const options = {
    attributes: [
      'id',
      'firstName',
      'lastName',
      'email',
      'image',
      'bio',
      'username',
    ],
  };

  if (req.query.search) {
    options.where = {
      where: {
        [Op.or]: [
          {
            username: {
              [Op.like]: `%${req.query.search}%`,
            },
          },
          {
            firstName: {
              [Op.like]: `%${req.query.search}%`,
            },
          },
          {
            lastName: {
              [Op.like]: `%${req.query.search}%`,
            },
          },
        ],
      },
    };
  } else {
    options.where = {
      isVerified: true,
    };
  }


  const users = await Users.findAll({ options });

  return users;
};
```

File: src/middlewares/Authorization.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+   * check if user is authourized
+   * @param {object} req - Request object
+   * @param {object} res - Response object
+   * @param {callback} next - The callback that passes the request
+   * @returns {null} - returns object
+   */
+  static async hasWriteAccess(req, res, next) {
+    if ((req.decoded.id !== req.body.userId) || !req.body.userId) {
+      return res.status(401).send({
+        code: 401,
```
Comment (kingisaac95) [inline src/middlewares/Authorization.js]: `success` key??

File: src/middlewares/Authorization.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+   * to the next handler
+   * @returns {object} res - Response object containing an error due
+   * to unauthorized access
+   */
+  static async checkToken(req, res, next) {
+    const token = req.body.token || req.query.token
+    || req.headers['x-access-token'];
+    if (!token) {
+      res.status(401).send({
+        code: 401,
```
Comment (kingisaac95) [inline src/middlewares/Authorization.js]: Ditto!

File: src/middlewares/Authorization.js [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+        message: 'User not authorized',
+      });
+    } else {
+      try {
+        const tokenVerified = Auth.verifyToken(token);
+        req.decoded = tokenVerified;
+        next();
+      } catch (error) {
+        res.status(401).send({
+          code: 401,
```
Comment (kingisaac95) [inline src/middlewares/Authorization.js]: Ditto!

File: src/routes/index.js [hunk 8]
```diff
@@ -25,9 +26,13 @@ const routes = (app) => {
   app.post(
     '/api/v1/articles',
     validateUserInputs.validateCreateArticle,
-    authentication.checkToken,
+    Authorization.checkToken,
```
Comment (kingisaac95) [inline src/routes/index.js:29]: Swap with line 29

File: src/routes/index.js [hunk 9]
```diff
@@ -25,9 +26,13 @@ const routes = (app) => {
...
     ArticleController.createArticle
   );
+  app.get(
+    '/api/v1/authors',
+    UserController.getAuthors
```
Comment (kingisaac95) [inline src/routes/index.js:36]: No token validation?

**Review by kingisaac95 — CHANGES_REQUESTED** [review]

**Review by kingisaac95 — APPROVED** [review]

**Pull request merged**